src/app/modules/myinfo/myinfo.adapter.ts

```diff
@@ -0,0 +1,276 @@
+  }
+
+  getUinFin(): string {
+    return this.#uinFin
+  }
+
+  _formatFieldValue(attr: ExternalAttr): string | undefined {
+    switch (attr) {
+      case ExternalAttr.MobileNo:
+        return formatPhoneNumber(this.#personData[attr])
+      case ExternalAttr.RegisteredAddress:
+        return formatAddress(this.#personData[attr])
+      case ExternalAttr.Vehicles:
+        return formatVehicleNumbers(this.#personData[attr])
+      case ExternalAttr.Occupation:
+        return formatOccupation(this.#personData[attr])
+      // Where field has code and description, return description
+      case ExternalAttr.Sex:
+      case ExternalAttr.Race:
+      case ExternalAttr.Dialect:
+      case ExternalAttr.Nationality:
+      case ExternalAttr.BirthCountry:
+      case ExternalAttr.ResidentialStatus:
+      case ExternalAttr.HousingType:
+      case ExternalAttr.HDBType:
+      case ExternalAttr.MaritalStatus:
+      case ExternalAttr.CountryOfMarriage:
+        return formatDescriptionField(this.#personData[attr])
+      // Remaining fields should only have 'value' key
+      case ExternalAttr.Name:
+      case ExternalAttr.PassportNumber:
+      case ExternalAttr.Employment:
+      case ExternalAttr.MarriageCertNumber:
+      case ExternalAttr.PassStatus:
+      case ExternalAttr.DateOfBirth:
+      case ExternalAttr.PassportExpiryDate:
+      case ExternalAttr.MarriageDate:
+      case ExternalAttr.DivorceDate:
+      case ExternalAttr.PassExpiryDate:
+        return formatBasicField(this.#personData[attr])
+      // Above cases should be exhaustive. Fall back to undefined
+      // as data shape is unknown.
+      default:
+        return undefined
+    }
+  }
+
+  /**
+   * Determine if frontend should lock the field to prevent it from being
+   * editable. The field is locked if it is government-verified and if it
+   * does not contain marriage-related information (decision by SNDGO & MSF due to
+   * overseas unregistered marriages). An empty myInfo field will always evaluate
+   * to false so that the field can be filled by form-filler.
+   *
+   * The affected marriage fields are:
+   * - marital
+   * - marriagedate
+   * - divorcedate
+   * - countryofmarriage
+   * - marriagecertno
+   *
+   * The function also uses the provided "source" flag within each MyInfo field to
+   * determine whether data is government verified.
+   *
+   * The mapping for "source" field is:
+   *
+   * 1 - Government-verified Data
+   * 2 - User Provided Data
+   * 3 - Not Applicable (e.g. CPF data for foreigners)
+   * 4 - Data retrieved from SingPass (e.g. email, mobileno)
+   *
+   */
+  _isDataReadOnly(
+    attr: ExternalAttr,
+    myInfoValue: string | undefined,
+  ): boolean {
+    const data = this.#personData[attr]
+    if (!data || !myInfoValue) return false
+    // Edge case: data is in array format
+    if (Array.isArray(data)) {
+      // All array items have source attribute
+      return (data as { source: MyInfoSource }[]).every(
```

Comment:
I'm curious what it would take to remove the need for this typecast here.

`data` is of type `MyInfoField<HDBOwnershipCustomFields>[] | ChildRecord[] | SponsoredChildRecord[] | MyInfoField<MyInfoVehicleCustomFields>[]`. All of them are arrays so in theory this should not need unsafe typecasting.